src/main/java/no/unit/nva/doi/transformer/CrossRefConverter.java

```diff
@@ -36,23 +55,125 @@
-                                                   .withMainTitle(extractTitle(document))
-                                                   .withPublicationType(extractPublicationType(document)).build())
-                                            .build();
+            return new Publication.Builder()
+                .withCreatedDate(now)
+                .withModifiedDate(now)
+                .withPublishedDate(createPublishedDate())
+                .withOwner(owner)
+                .withIdentifier(identifier)
+                .withPublisher(toPublisher(publisherId))
+                .withStatus(DEFAULT_NEW_PUBLICATION_STATUS)
+                .withIndexedDate(createIndexedDate())
+                .withHandle(createHandle())
+                .withLink(createLink())
+                .withProject(createProject())
+                .withLicense(createLicense())
+                .withFileSet(createFilseSet())
+                .withEntityDescription(new EntityDescription.Builder()
+                    .withContributors(toContributors(document.getAuthor()))
+                    .withDate(extractDate(document).orElse(null))
+                    .withMainTitle(extractTitle(document))
+                    .withAlternativeTitles(extractAlternativeTitles(document))
+                    .withPublicationType(extractPublicationType(document))
+                    .withPublicationSubtype(extractPublicationSubtype())
+                    .withAbstract(extractAbstract(document))
+                    .withLanguage(extractLanguage(document))
+                    .withNpiSubjectHeading(extractNpiSubjectHeading())
+                    .withTags(extractTags())
+                    .withDescription(extractDescription())
+                    .withReference(extractReference(document))
+                    .withMetadataSource(extractMetadataSource(document))
+                    .build())
+                .build();
         }
         throw new IllegalArgumentException(INVALID_ENTRY_ERROR);
     }
 
+    private URI extractMetadataSource(CrossRefDocument document) {
+        if (containsCrossrefAsSource(document)) {
+            return CROSSEF_URI;
+        } else {
+            return tryCreatingUri(document.getSource()).orElse(null);
+        }
+    }
+
+    private boolean containsCrossrefAsSource(CrossRefDocument document) {
+        return Optional.ofNullable(document.getSource())
+                       .map(str -> str.toLowerCase(Locale.getDefault()))
+                       .filter(str -> str.contains(CROSSREF))
+                       .isPresent();
+    }
+
+    private Optional<URI> tryCreatingUri(String source) {
+        try {
+            return Optional.of(URI.create(source));
+        } catch (Exception e) {
+            return Optional.empty();
+        }
+    }
+
+    private Reference extractReference(CrossRefDocument document) {
+        PublicationContext context = new PublicationContext.Builder()
+            .withLevel(null)
```

Comment:
I suppose this is the "Norsk vitenskapsindeks - NVI" and crossref does not give us any info about it. I am putting null just to be explicit of what the value will be. It will be null, even if I do not set anything because it is the default value. 
I am not sure what is best, to keep it, or remove it.